pkg/client/oci/oci.go

```diff
@@ -0,0 +1,302 @@
+
+// ListChartVersions lists all versions of a chart
+func (r *Repo) ListChartVersions(name string) ([]string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), getTimeout)
+	defer cancel()
+
+	u := *r.url
+	// Form API endpoint URL from repository URL
+	u.Path = path.Join("v2", u.Path, name, "tags", "list")
+
+	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if r.username != "" && r.password != "" {
+		req.SetBasicAuth(r.username, r.password)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer resp.Body.Close()
+
+	status := resp.StatusCode
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Errorf("unexpected response — %d %q — from %s", status, http.StatusText(status), u.String())
+	}
+	// If status not found 404, it could mean the asset has no release yet
+	if status == http.StatusNotFound {
+		return []string{}, nil
+	}
+	if status != http.StatusOK {
+		return nil, errors.Errorf("unexpected response — %d %q, %s — from %s", status, http.StatusText(status), string(body), u.String())
+	}
+
+	ot := &Tags{}
+	if err := json.Unmarshal(body, ot); err != nil {
+		return nil, err
+	}
+	chartTags := []string{}
+	for _, tag := range ot.Tags {
+		tm, err := r.getTagManifest(name, tag)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if tm.Config.MediaType == "application/vnd.cncf.helm.config.v1+json" {
+			chartTags = append(chartTags, tag)
+		} else {
+			klog.V(4).Infof("Skipping %q tag as it is not chart type", tag)
+		}
+	}
+	return chartTags, nil
+}
+
+// GetDownloadURL returns the URL to download a chart
+func (r *Repo) GetDownloadURL(name string, version string) (string, error) {
+	digest, err := r.getChartDigest(name, version)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	u := *r.url
+	// Form API endpoint URL from repository URL
+	u.Path = path.Join("v2", u.Path, name, "blobs", digest)
+	return u.String(), nil
+}
+
+// Fetch fetches a chart
+func (r *Repo) Fetch(filename string, name string, version string) error {
+	u, err := r.GetDownloadURL(name, version)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if r.username != "" && r.password != "" {
+		req.SetBasicAuth(r.username, r.password)
+	}
+
+	reqID := utils.EncodeSha1(u + filename)
+	klog.V(4).Infof("[%s] GET %q", reqID, u)
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if err != nil {
+		return errors.Annotatef(err, "fetching %s:%s chart", name, version)
+	}
+	defer res.Body.Close()
+
+	if ok := res.StatusCode >= 200 && res.StatusCode <= 299; !ok {
+		bodyStr := utils.HTTPResponseBody(res)
+		return errors.Errorf("unable to fetch %s:%s chart, got HTTP Status: %s, Resp: %v", name, version, res.Status, bodyStr)
+	}
+	klog.V(4).Infof("[%s] Got HTTP Status: %s", reqID, res.Status)
```

Comment:
let's use a `switch` and check the specific response codes that the API may return

Reply:
I just need to check for StatusOK, the rest of them should return an error. Do you think it still makes sense the switch?

Reply:
yes, so it is simpler to add additional behaviors in the future and it is consistent with the other method